File: Java/15-3Sum.java
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int len = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length < 3)    return res;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length-2; i++){
            if (i > 0 && nums[i] == nums[i-1])    
                continue; //skip same results
            int target = 0 - nums[i];
            int j = i + 1, k = len - 1;
            while (j < k){
                if (nums[j] + nums[k] == target){
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while (j < k && nums[j] == nums[j+1])   
                        j++;//skip same results
                    while (j < k && nums[k] == nums[k-1])   
                        k--;//skip same results
                    j++;
                    k--;
                }else if (nums[j] + nums[k] < target)  
                    j++;
                else 
                    k--;
            }
        }
        return res;
    }
}

File: cpp/15-3Sum.cpp
//[-1, 0, 1, 2, -1, -4]
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        if (nums.size() < 3)
            return res;
        int first = 0, second = 1, third = nums.size() - 1;
        for (; first < nums.size() - 2; first++) {
            if (first > 0 && nums[first] == nums[first-1]) {
                continue;
            }
            second = first + 1;
            third = nums.size() - 1;
            while (second < third) {
                if (nums[first] + nums[second] + nums[third] == 0) {
                    res.push_back({nums[first], nums[second], nums[third]});
                    while (second < third && nums[second] == nums[second+1]) {
                        second++;
                    }
                    while (second < third && nums[third] == nums[third-1]) {
                        third--;
                    }
                    second++;
                    third--;
                } else if (nums[first] + nums[second] + nums[third] > 0) {
                    third--;
                } else {
                    second++;
                }
            }
        }
        return res;
    }
};
